### DISCUSSION TITLE
Construct a temporary object without copying

### DISCUSSION THREAD

**Discussion by metab0t** [wjakob/nanobind#614] [General]
In my optimization modeling language, I expose some expression class to Python like
```cpp
struct Expr{
  ...
  ...
};
```

```Python
expr = Expr(5)
expr += x
```

What I want is to construct a temporary constraint object when I write `expr<=1.0` and avoids copying the `expr` object while keeping memory safe.

I assume that `Constraint` contains a pointer to the `Expr` like
```cpp
struct Constraint{
  Expr* expr;
  char sense;
  double rhs;
};

Constraint operator<=(const Expr& expr, double rhs){
  return Constraint{&expr, '<', rhs};
}
```

But the code above seems not memory safe to store the raw pointer directly for the following Python code:
```Python
expr = Expr(5)
expr += x
constraint = expr <= 0.0
```

Is there some good solutions to this problem?

**Comment by wjakob** [comment]
You can use ``keep_alive`` and return value policies to do some of these things, but you have to be very careful to get things right. Please see the documentation.

The safer standard solution is to adopt either STL shared pointers or a custom intrusive reference counting-based approach. Again, both are extensively discussed in the documentation.
